nsxt/metadata/metadata.go

```diff
@@ -0,0 +1,275 @@
+		}
+	}
+}
+
+// SchemaToStruct converts terraform schema to NSX model struct
+// currently supports nested subtype and trivial types
+func SchemaToStruct(elem reflect.Value, d *schema.ResourceData, metadata map[string]*ExtendedSchema, parent string, parentMap map[string]interface{}) {
+	for key, item := range metadata {
+		if item.Metadata.ReadOnly {
+			log.Printf("[INFO] skip key %s as read only", key)
+			continue
+		}
+		if item.Metadata.Skip {
+			log.Printf("[INFO] skip key %s", key)
+			continue
+		}
+		if item.Metadata.IntroducedInVersion != "" && util.NsxVersionLower(item.Metadata.IntroducedInVersion) {
+			log.Printf("[INFO] skip key %s as NSX does not have support", key)
+			continue
+		}
+
+		log.Printf("[INFO] inspecting key %s with type %s", key, item.Metadata.SchemaType)
+		if len(parent) > 0 {
+			log.Printf("[INFO] parent %s key %s", parent, key)
+		}
+		if item.Metadata.SchemaType == "string" {
+			var value string
+			if len(parent) > 0 {
+				value = parentMap[key].(string)
+			} else {
+				value = d.Get(key).(string)
+			}
+			log.Printf("[INFO] assigning string %v to %s", value, key)
+			elem.FieldByName(item.Metadata.SdkFieldName).Set(reflect.ValueOf(&value))
+		}
+		if item.Metadata.SchemaType == "bool" {
+			var value bool
+			if len(parent) > 0 {
+				value = parentMap[key].(bool)
+			} else {
+				value = d.Get(key).(bool)
+			}
+			log.Printf("[INFO] assigning bool %v to %s", value, key)
+			elem.FieldByName(item.Metadata.SdkFieldName).Set(reflect.ValueOf(&value))
+		}
+		if item.Metadata.SchemaType == "int" {
+			var value int64
+			if len(parent) > 0 {
+				value = int64(parentMap[key].(int))
+			} else {
+				value = int64(d.Get(key).(int))
+			}
+			log.Printf("[INFO] assigning int %v to %s", value, key)
+			elem.FieldByName(item.Metadata.SdkFieldName).Set(reflect.ValueOf(&value))
+		}
+		if item.Metadata.SchemaType == "struct" {
+			nestedObj := reflect.New(item.Metadata.ReflectType)
+			var itemList []interface{}
+			if len(parent) > 0 {
+				itemList = parentMap[key].([]interface{})
+			} else {
+				itemList = d.Get(key).([]interface{})
+			}
+			if len(itemList) == 0 {
+				continue
+			}
+			nestedSchema := itemList[0].(map[string]interface{})
+
+			childElem := item.Schema.Elem.(*ExtendedResource)
+			SchemaToStruct(nestedObj.Elem(), d, childElem.Schema, key, nestedSchema)
+			log.Printf("[INFO] assigning struct %v to %s", nestedObj, key)
+			elem.FieldByName(item.Metadata.SdkFieldName).Set(nestedObj)
+		}
+		if item.Metadata.SchemaType == "list" || item.Metadata.SchemaType == "set" {
+			var itemList []interface{}
+			if item.Metadata.SchemaType == "list" {
+				if len(parent) > 0 {
+					itemList = parentMap[key].([]interface{})
+				} else {
+					itemList = d.Get(key).([]interface{})
+				}
+			} else {
+				if len(parent) > 0 {
+					itemList = parentMap[key].(*schema.Set).List()
+				} else {
+					itemList = d.Get(key).(*schema.Set).List()
+				}
+			}
+
+			if len(itemList) == 0 {
+				continue
```

Comment:
I wonder if here we might run into a problem of assigning `nil` value to array instead of assigning empty array?
We need to be able to remove a list of values from a struct, and in some cases `nil` value would be ignored by the Patch call. 
We can leave this as is for now, and consider adding another metadata value that would control this behavior when the need arises.

Reply:
Nice catch. I actually think there is no harm in passing an empty slice if tf schema is empty. For create this is a no op, and for update this is an explict erasure. We can revisit this part if other handling for erasing is needed for some NSX resources.